src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -147,60 +156,103 @@ TryToDelegateFunctionCall(DistributedPlanningContext *planContext)
 		return NULL;
 	}
 
+	FuncExpr *fromFuncExpr = NULL;
 	if (joinTree->fromlist != NIL)
 	{
+		if (list_length(joinTree->fromlist) != 1)
+		{
+			/* e.g. SELECT ... FROM rel1, rel2. */
+			Assert(list_length(joinTree->fromlist) > 1);
+			return NULL;
+		}
+
 		/*
 		 * In pg12's planning phase empty FROMs are represented with an RTE_RESULT.
 		 * When we arrive here, standard_planner has already been called which calls
 		 * replace_empty_jointree() which replaces empty fromlist with a list of
 		 * single RTE_RESULT RangleTableRef node.
 		 */
-		if (list_length(joinTree->fromlist) == 1)
+		RangeTblRef *reference = linitial(joinTree->fromlist);
+
+		if (IsA(reference, RangeTblRef))
 		{
-			RangeTblRef *reference = linitial(joinTree->fromlist);
+			RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
+											  planContext->query->rtable);
 
-			if (IsA(reference, RangeTblRef))
-			{
-				RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
-												  planContext->query->rtable);
-				if (rtentry->rtekind != RTE_RESULT)
-				{
-					/* e.g. SELECT f() FROM rel */
-					return NULL;
-				}
-			}
-			else
+			if (rtentry->rtekind == RTE_FUNCTION)
 			{
 				/*
-				 * e.g. IsA(reference, JoinExpr). This is explicit join expressions
-				 * like INNER JOIN, NATURAL JOIN, ...
+				 * Look for a function in the FROM clause.
 				 */
+				fromFuncExpr = FunctionInFromClause(joinTree->fromlist,
+													planContext->query);
+			}
+			else if (rtentry->rtekind != RTE_RESULT)
+			{
+				/* e.g. SELECT f() FROM rel */
+				ereport(DEBUG4, (errmsg("FromList item is not empty")));
 				return NULL;
 			}
 		}
 		else
 		{
-			/* e.g. SELECT ... FROM rel1, rel2. */
-			Assert(list_length(joinTree->fromlist) > 1);
+			/*
+			 * e.g. IsA(reference, JoinExpr). This is explicit join expressions
+			 * like INNER JOIN, NATURAL JOIN, ...
+			 */
 			return NULL;
 		}
 	}
 
+	FuncExpr *targetFuncExpr = NULL;
 	List *targetList = planContext->query->targetList;
-	if (list_length(planContext->query->targetList) != 1)
+	int targetListLen = list_length(targetList);
+
+	if (targetListLen > 1)
 	{
 		/* multiple target list items */
 		return NULL;
```

Comment:
it would be a bit nicer if this check did not apply when the function call is in FROM to allow `SELECT a, b FROM force_delegated_call(3)`

Reply:
Done.